examples/generative/gaugan.py

```diff
@@ -0,0 +1,792 @@
+        )
+        self.discriminator_loss = DiscriminatorLoss()
+        self.feature_matching_loss = FeatureMatchingLoss()
+        self.vgg_loss = VGGFeatureMatchingLoss()
+
+    def train_discriminator(self, latent_vector, segmentation_map, real_image, labels):
+        fake_images = self.generator([latent_vector, labels])
+        with tf.GradientTape() as gradient_tape:
+            pred_fake = self.discriminator([segmentation_map, fake_images])[-1]
+            pred_real = self.discriminator([segmentation_map, real_image])[-1]
+            loss_fake = self.discriminator_loss(pred_fake, False)
+            loss_real = self.discriminator_loss(pred_real, True)
+            total_loss = 0.5 * (loss_fake + loss_real)
+
+        self.discriminator.trainable = True
+        gradients = gradient_tape.gradient(
+            total_loss, self.discriminator.trainable_variables
+        )
+        self.discriminator_optimizer.apply_gradients(
+            zip(gradients, self.discriminator.trainable_variables)
+        )
+        return total_loss
+
+    def train_generator(
+        self, latent_vector, segmentation_map, labels, image, mean, variance
+    ):
+        # Generator learns through the signal provided by the discriminator. During
+        # backpropagation, we only update the generator parameters.
+        self.discriminator.trainable = False
+        with tf.GradientTape() as tape:
+            real_d_output = self.discriminator([segmentation_map, image])
+            fake_d_output, fake_image = self.combined_model(
+                [latent_vector, labels, segmentation_map]
+            )
+            pred = fake_d_output[-1]
+
+            # Compute generator losses.
+            g_loss = generator_loss(pred)
+            kl_loss = self.kl_divergence_loss_coeff * kl_divergence_loss(mean, variance)
+            vgg_loss = self.vgg_feature_loss_coeff * self.vgg_loss(image, fake_image)
+            feature_loss = self.feature_loss_coeff * self.feature_matching_loss(
+                real_d_output, fake_d_output
+            )
+            total_loss = g_loss + kl_loss + vgg_loss + feature_loss
+
+        gradients = tape.gradient(total_loss, self.combined_model.trainable_variables)
+        self.generator_optimizer.apply_gradients(
+            zip(gradients, self.combined_model.trainable_variables)
+        )
+        return total_loss, feature_loss, vgg_loss, kl_loss
+
+    def train_step(self, data):
+        segmentation_map, image, labels = data
+        mean, variance = self.encoder(image)
+        latent_vector = self.sampler([mean, variance])
+        discriminator_loss = self.train_discriminator(
+            latent_vector, segmentation_map, image, labels
+        )
+        (generator_loss, feature_loss, vgg_loss, kl_loss) = self.train_generator(
+            latent_vector, segmentation_map, labels, image, mean, variance
+        )
+        return {
+            "discriminator_loss": discriminator_loss,
```

Comment:
This will only display the loss value for the latest batch rather than the average over the batches of the current epoch. To do the latter, use a `Mean()` metric tracker listed in `self.metrics` and `return metric.result() for metric in self.metrics` after updating the trackers above